fix(tab): tabpanel is tabbable when content has no tabbable content

tay1orjones opened the pull request:
Closes #7078

~Add `hasNoFocusableContent` prop to Tab to inform if the tabpanel container should be focusable. This preserves existing functionality and allows users to opt-in to the new focus behavior when their tab content is text only or has no focusable elements/content.~

Add `useTabbableContent` to TabContent to automatically determine if the contents contains a tabbable element.

#### Changelog

**New**

- Add `useTabbableContent` hook to TabContent

**Changed**

- Update Tabs story to have a tabbable button in the contents of one tab
- Refactor TabContent tests to use testing-library and test for tabindex/hook functionality

#### Testing / Reviewing

1. Open the [default Tabs story](https://deploy-preview-7894--carbon-components-react.netlify.app/?path=/story/components-tabs--default)
1. Type the <kbd>Tab</kbd> key - focus should move to the tabpanel/content container.
1. Type <kbd>Shift + Tab</kbd> - focus moves back to "Tab Label 1"
1. Type <kbd>→</kbd> - focus should move to "Tab Label 2"
1. Type the <kbd>Tab</kbd> key again - focus should move to the button inside the content (the tabpanel should not be focused as the content contains a focusable element)

tay1orjones committed:
fix(tab): tabpanel is focusable when content has no focusable content

joshblack reviewed (commented):
Would we want to do this automatically for users, potentially? Like in an effect. Something like:

```jsx
function TabContent(props) {
  // ...
  const ref = useRef(null);
  const hasFocusableContent = useFocusableContent(ref);

  // ...
  return <div role="tabpanel" tabIndex={hasFocusableContent ? 0 : undefined}>...</div>
}

function useFocusableContent(ref) {
  const [hasFocusableContent, setFocusableContent] = useState(false);

  useEffect(() => {
    const elements = getFocusableElements(ref.current);
    if (elements.length > 0) {
      setFocusableContent(true);
    }
  }, [ref]);

  return hasFocusableContent;
}
```

tay1orjones commented:
I dig the magic there for sure. It would also work if content is passed in via `renderContent`. 

My only concern would be performance of `getFocusableElements(ref.current);`. Iterating over every element recursively against a known list of focusable elements on every render feels ... expensive 😬

tay1orjones commented:
You know what... I forgot about `querySelectorAll` 😅  Yeah this sounds good, I'll update the PR.

emyarod commented:
this might be helpful for your querySelectorAll https://github.com/carbon-design-system/carbon/blob/master/packages/react/src/internal/keyboard/navigation.js#L64-L72

tay1orjones committed:
fix(tabcontent): automatically determine if contents is tabbable

tay1orjones committed:
Merge branch 'master' into 7078-tab-content-focus

tay1orjones commented:
Thanks @emyarod, I hadn't seen that. I think it's `selectorTabbable` that I really wanted, "focusable elements" was a misnomer. 

I modified the PR body to reflect the new approach. Also added some tests and refactored the tests off enzyme.

emyarod commented on packages/react/src/components/TabContent/TabContent.js:
```diff
@@ -6,24 +6,46 @@
+      // we don't need to know which element or how many are tabbable, just that there is _something_ tabbable
+      node.querySelector(selectorTabbable)
+        ? setHasTabbableContent(true)
+        : setHasTabbableContent(false);
```
```suggestion
      setHasTabbableContent(node.querySelector(selectorTabbable));
```

maybe something like this? seems simpler than a ternary statement

tay1orjones committed:
Update packages/react/src/components/TabContent/TabContent.js

Co-authored-by: emyarod <emyarod@users.noreply.github.com>

jnm2377 reviewed (approved):
LGTM! Tab content is now tabbable if there's no interactive elements. Thanks for adding the comments to the code too! Super helpful! 🔥

joshblack commented:
@tay1orjones definitely get what you mean, it might be that this abstraction doesn't make see if the tabpanel also needs to be explicitly labeled so would totally understand moving forward with the prop approach if this wouldn't be a good fit to do automatically.

tay1orjones commented:
@joshblack It works quite nicely. Explicitly labelling the tabpanel feels like a very unlikely use case, but we could allow it by placing the automatic tabindex above `{...others}` so it's overridable?

joshblack commented on packages/react/src/components/TabContent/TabContent.js:
```diff
@@ -6,24 +6,44 @@
   const tabContentClasses = classNames(`${prefix}--tab-content`, {
     [className]: className,
   });
+  const [hasTabbableContent, ref] = useTabbableContent();
```
Love this callback pattern, seems super clever for running code when a component "mounts". I wonder if there would there be an easy way to have this hook accept a ref just to make composition with hooks that use refs a little easier?

Something like:

```js
function useTabbableContent(ref) {
  const [hasTabbableContent, setHasTabbableContent] = useState(false);

  // useEffect on the server, useLayoutEffect on the client
  useIsomorphicEffect(() => {
    if (ref.current) {
      setHasTabbableContent(ref.current.querySelector(selectorTabbable));
    }
  });

  return hasTabbableContent;
}
```

I _think_ these would run at similar times but wasn't sure 🤷

tay1orjones replied:
Great idea! Updated

tay1orjones committed:
Merge branch 'main' of github.com:carbon-design-system/carbon into 7078-tab-content-focus

tay1orjones committed:
feat(tabs): add useIsomorphicEffect, reconfigure useTabbableContent

tay1orjones committed:
Merge branch 'main' into 7078-tab-content-focus

tay1orjones committed:
Merge branch 'main' into 7078-tab-content-focus

tay1orjones committed:
Merge branch 'main' into 7078-tab-content-focus

andreancardona reviewed (approved):
@tay1orjones  this looks awesome!

I'm so so sorry for the review delay... 👀

tay1orjones committed:
Merge branch 'main' into 7078-tab-content-focus

tay1orjones committed:
Merge branch 'main' into 7078-tab-content-focus

kodiakhq[bot] committed:
Merge branch 'main' into 7078-tab-content-focus
